File: Python方案下载WordPress文章到Markdown/downloadWordpresstoMarkdown.py
import os
import re
import requests
import html2text
import yaml

# WordPress REST API 地址（请根据实际站点修改）
API_URL = "https://csrwiki.com/wp-json/wp/v2"
# 保存 Markdown 文件的目录
SAVE_DIR = "./exported_posts"

# 确保保存目录存在
os.makedirs(SAVE_DIR, exist_ok=True)


def get_all(endpoint):
    """
    从 WordPress REST API 分页获取数据
    endpoint: API 子路径，例如 "posts", "categories", "tags", "users", "media"
    return: 返回所有数据的列表
    """
    items = []
    page = 1
    while True:
        r = requests.get(f"{API_URL}/{endpoint}", params={"per_page": 100, "page": page})
        if r.status_code != 200:
            print(f"⚠️ 请求失败: {endpoint}, 状态码 {r.status_code}")
            break
        data = r.json()
        if not data:  # 没有更多数据，退出循环
            break
        items.extend(data)
        page += 1
    return items


def sanitize_filename(filename: str) -> str:
    """
    清理文件名中不合法的字符，避免保存失败
    """
    return re.sub(r'[\\/*?:"<>|]', "_", filename)


def main():
    print("🔍 获取分类和标签...")
    categories = {c["id"]: c["name"] for c in get_all("categories")}
    tags = {t["id"]: t["name"] for t in get_all("tags")}

    print("🔍 获取作者信息...")
    authors = {a["id"]: a["name"] for a in get_all("users")}

    print("🔍 获取文章...")
    posts = get_all("posts")

    print("🔍 获取媒体（特色图片）...")
    media = {m["id"]: m["source_url"] for m in get_all("media")}

    # 初始化 HTML → Markdown 转换器
    h = html2text.HTML2Text()
    h.ignore_links = False  # 保留超链接

    for post in posts:
        post_id = post["id"]
        title = post["title"]["rendered"]
        date = post["date"]
        content_html = post["content"]["rendered"]

        # 转换 HTML → Markdown
        content_md = h.handle(content_html)

        # 分类和标签
        post_categories = [categories.get(cid, "未分类") for cid in post["categories"]]
        post_tags = [tags.get(tid, f"tag-{tid}") for tid in post["tags"]]

        # 作者信息
        author_name = authors.get(post["author"], "未知作者")

        # 特色图片（如果有）
        featured_image = None
        if post.get("featured_media"):
            featured_image = media.get(post["featured_media"], None)

        # front matter
        front_matter = {
            "id": post_id,
            "title": title,
            "date": date,
            "author": author_name,
            "categories": post_categories,
            "tags": post_tags
        }
        if featured_image:
            front_matter["featured_image"] = featured_image

        # 组合 front matter 和正文
        md_content = f"---\n{yaml.dump(front_matter, allow_unicode=True)}---\n\n{content_md}"

        # 用中文标题作为文件名
        safe_title = sanitize_filename(title)
        filename = os.path.join(SAVE_DIR, f"{safe_title}.md")

        with open(filename, "w", encoding="utf-8") as f:
            f.write(md_content)

        print(f"✅ 已保存: {filename}")


if __name__ == "__main__":
    main()

File: NodeJs方案下载英文WordPress文章到Markdown/downloadWordpressToMarkdown.js
import fs from "fs";
import path from "path";
import { unified } from "unified";
import rehypeParse from "rehype-parse";
import rehypeRaw from "rehype-raw";
import rehypeRemark from "rehype-remark";
import remarkGfm from "remark-gfm";
import remarkStringify from "remark-stringify";
import yaml from "js-yaml";

const API_URL = "https://csrwiki.com/wp-json/wp/v2";
// Polylang 语言代码（仅抓取指定语言内容）
const LANG = "en"; // 如需切换，请改为 "zh" 等
const SAVE_DIR = "./exported_posts";

fs.mkdirSync(SAVE_DIR, { recursive: true });

/**
 * 分页抓取 WordPress API
 */
async function getAll(endpoint, query = {}) {
  let page = 1;
  const items = [];
  while (true) {
    const searchParams = new URLSearchParams({ per_page: "100", page: String(page) });
    for (const [k, v] of Object.entries(query)) {
      if (v !== undefined && v !== null) searchParams.append(k, String(v));
    }
    const res = await fetch(`${API_URL}/${endpoint}?${searchParams.toString()}`);
    if (!res.ok) {
      if (res.status === 400) {
        // 400 说明已到最后一页，正常结束循环
        break;
      }
      console.warn(`⚠️ 请求失败: ${endpoint}, 状态码 ${res.status}`);
      break;
    }
    const data = await res.json();
    if (!data.length) break;
    items.push(...data);
    page++;
  }
  return items;
}

/**
 * HTML → Markdown
 */
async function htmlToMarkdown(html) {
  const file = await unified()
    .use(rehypeParse, { fragment: true })
    .use(rehypeRaw) // 支持混合 HTML
    .use(rehypeRemark) // HTML → Markdown AST
    .use(remarkGfm) // 支持 GFM (表格/任务列表等)
    .use(remarkStringify, {
      fences: true,
      bullet: "-",
      entities: "escape",
    })
    .process(html);
  return String(file);
}

/**
 * 清理文件名
 */
function sanitizeFilename(filename) {
  return filename.replace(/[\\/*?:"<>|]/g, "_");
}

/**
 * 主逻辑
 */
async function main() {
  console.log("🔍 获取分类和标签...");
  const categoriesArr = await getAll("categories", { lang: LANG });
  const categories = {};
  categoriesArr.forEach((c) => (categories[c.id] = c.slug));

  const tagsArr = await getAll("tags", { lang: LANG });
  const tags = {};
  tagsArr.forEach((t) => (tags[t.id] = t.slug));

  // 去掉作者处理，不再请求 users

  console.log("🔍 获取文章...");
  const posts = await getAll("posts", { lang: LANG });

  console.log("🔍 获取媒体...");
  const mediaArr = await getAll("media", { lang: LANG });
  const media = {};
  mediaArr.forEach((m) => (media[m.id] = m.source_url));

  for (const post of posts) {
    const postId = post.id;
  const title = post.title.rendered || "untitled";
  const slug = post.slug || `post-${postId}`;
    const date = post.date;
    const contentHtml = post.content.rendered;

    const contentMd = await htmlToMarkdown(contentHtml);

    const postCategories = post.categories.map(
      (cid) => categories[cid] || "uncategorized"
    );
    const postTags = post.tags.map((tid) => tags[tid] || `tag-${tid}`);
  // 已移除作者逻辑

    let featuredImage = null;
    if (post.featured_media) {
      featuredImage = media[post.featured_media] || null;
    }

    // 文章URL
    const postUrl = post.link || '';

    const frontMatter = {
      id: postId,
      title: title,
      slug: slug,
      date: date,
      language: LANG,
      categories: postCategories,
      tags: postTags,
      url: postUrl,
    };
    if (featuredImage) {
      frontMatter.featured_image = featuredImage;
    }

    const mdContent =
      `---\n${yaml.dump(frontMatter, { skipInvalid: true, lineWidth: -1 })}---\n\n` +
      contentMd;

    // 以第一个分类为主目录，若无分类则放 uncategorized
    const mainCategory = postCategories[0] || "uncategorized";
    const categoryDir = path.join(SAVE_DIR, sanitizeFilename(mainCategory));
    fs.mkdirSync(categoryDir, { recursive: true });

    const safeTitle = sanitizeFilename(title);
    const filename = path.join(categoryDir, `${safeTitle}.md`);
    fs.writeFileSync(filename, mdContent, "utf-8");

    console.log(`✅ 已保存: ${filename}`);
  }
}

main();